---
title: "05. Nematostella vectensis metacell analysis"
date: "`r format(Sys.time(), '%d %B, %Y')`"
auhor: "Anamaria Elek"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: false
    toc_depth: 2
---
```{r options, include=FALSE}
knitr::opts_chunk$set(eval = FALSE, fig.align = 'center')
```

# Set up metacell analysis

Load required packages and functions.

```{r setup, eval=TRUE, message=FALSE, warning=FALSE}
library(data.table)
library(ggplot2)
library(ggrepel)
library(ggrastr)
library(plotly)
library(scales)
library(stringr)
library(RColorBrewer)
library(circlize)
library(tgconfig)
library(metacell)
theme_py <- theme_classic() + theme(
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA),
  text = element_text(size=20)
)
theme_set(theme_py)
# load functions for downstream analysis
scripts <- list.files("metacell_downstream_functions",full.names=TRUE)
for (file in scripts) source(file)
```

Initialize database and figures output directory.

```{r mc_setup, eval=TRUE, warning=FALSE, message=FALSE}
require(metacell)

# analysis dir
cluster_dir <- "clustering_nematostella"
dir.create(cluster_dir, showWarnings=FALSE)

# initialize db directory
scdb_dir <- file.path(cluster_dir, "scdb")
dir.create(scdb_dir, showWarnings=FALSE)
scdb_init(scdb_dir, force_reinit=TRUE)

# plotting directory
scdb_fig_dir <- file.path(scdb_dir,"figs")
dir.create(scdb_fig_dir, showWarnings=FALSE)
scfigs_init(scdb_fig_dir)

# directory with annotations
annot_dir <- "annotation_nematostella"
```

# Metacell piepline

## Input data filtering

Load the expression data to metacell database.  

```{r mcell_import}
# import UMI table
mat_nm <- "nvec"
mcell_import_multi_mars(
  mat_nm = mat_nm,
  dataset_table_fn = file.path(cluster_dir, "MARS_Batches.txt"),
  base_dir = "umi.tab",
  force = TRUE
)
mat <- scdb_mat(mat_nm)
```

Filter ERCCs

```{r mcell_filter_genes}
erccs <- rownames(mat@mat)[grepl("ERCC",rownames(mat@mat))] 
peaks_counts <- rowSums(as.matrix(mat@mat[grepl("_peak_",rownames(mat@mat)),]))
peaks_ignore <- names(which(peaks_counts<100))
mcell_mat_ignore_genes(new_mat_id = mat_nm, mat_id = mat_nm, ig_genes = c(erccs,peaks_ignore))
mat <- scdb_mat(mat_nm)
```

Filter by cell depth (UMIs/cell)

```{r mcell_filter}
x <- colSums(as.matrix(mat@mat))
mcell_plot_umis_per_cell(mat_id = mat_nm, min_umis_cutoff = 100)
large_cells <- names(which(x>10000))
small_cells <- names(which(x<100))
mat_nm_filt <- "nvec_100"
mcell_mat_ignore_cells(new_mat_id = mat_nm_filt ,mat_id = mat_nm, ig_cells=c(small_cells,large_cells))
mat <- scdb_mat(mat_nm_filt)
```

Calculate gene statistics

```{r mcell_gstat}
mcell_add_gene_stat(gstat_id = "gstat", mat_id = mat_nm_filt, force = TRUE)
mat <- scdb_mat(mat_nm_filt)
```

## Building metacells

Select marker genes for building metacell graph.  

```{r mcell_markers}
morange <- rownames(mat@mat)[grepl("mOrange",rownames(mat@mat))] 
bkg <- readLines(file.path(annot_dir,"black_list_genes"))
mcell_gset_filter_multi(
  gstat_id = "gstat",
  gset_id = "clust_markers",
  T_tot = 30, T_top3 = 2, T_szcor = -0.05, T_niche = 0.05, 
  blacklist = c(morange,bkg), force_new=TRUE
)
mcell_plot_gstats(gstat_id = "gstat", gset_id = "clust_markers")
```

Build metacells

```{r mcell_knn}
# build cell graph
mcell_add_cgraph_from_mat_bknn(
  mat_id = mat_nm_filt,
  gset_id = "clust_markers",
  graph_id = "graphk100", K = 100, dsamp=FALSE
)

# resample graph and compute metacells for each resampling
mcell_coclust_from_graph_resamp(
  coc_id = "coc500_min20",
  graph_id = "graphk100",
  min_mc_size = 20, p_resamp = 0.75, n_resamp = 1000
)

# final graph and metacells from co-occurence in resampled graphs
mcell_mc_from_coclust_balanced(
  mc_id = "mc_k30",
  coc_id = "coc500_min20",
  mat_id = mat_nm_filt,
  K = 30, min_mc_size = 20, alpha = 2 # changed from K=100
)

mc <- scdb_mc("mc_k30")
```

## Filtering bad metacells

```{r mcell_load_filt, echo=FALSE, eval=TRUE}
mc_id <- "mc_k30"
mat_id <- "nvec_100"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
mcs <- colnames(mc@mc_fp)
marker_genes_gset <- scdb_gset(id="clust_markers")
marker_genes <- names(marker_genes_gset@gene_set)
```

The following criteria are used for filtering out bad metacells:

* **number of genes with non-zero UMIs** summarized across all cells in metacell - metacells consisting of cells that express very low number of genes are potentially bad metacells

```{r nonzero_umi}
# count genes with non-zero UMIs in single cells
csums <- lapply(mcs, function(mcid) {
  mc_cells <- names(mc@mc[mc@mc==mcid])
  mc_cells <- intersect(mc_cells,colnames(mat@mat))
  nonzerogenes <- Matrix::colSums(mat@mat[marker_genes,mc_cells]>0)
  data.table(non_zero_genes_per_cell = nonzerogenes)
})
csumsdt <- rbindlist(csums,idcol = "mc")
csumsdt[,mc:=factor(mc,levels=mcs)]
# summarize by taking the peak of distribution of non-zero UMI counts in the cells in each metacell
csumsdtmean <- csumsdt[,.SD[,.(
  non_zero_genes_peak={
    dns = density(non_zero_genes_per_cell)
    ind = which.max(dns$y)
    dns$x[ind]
})],mc]
```

* **total number of UMIs** summed across all cells in metacell - metacells consisting of very small cells, with a low UMI count, are potentially bed metacells.

```{r total_umi}
# count total UMIs in cells
umisums <- lapply(mcs, function(mcid) {
  mc_cells <- names(mc@mc[mc@mc==mcid])
  mc_cells <- intersect(mc_cells,colnames(mat@mat))
  mc_table <- mat@mat[marker_genes,mc_cells]
  totalumis <- Matrix::colSums(mc_table)
  data.table(total_umis_per_cell = totalumis)
})
umisumsdt <- rbindlist(umisums,idcol = "mc")
umisumsdt[,mc:=factor(mc,levels=mcs)]
# summarize by taking the sum of total UMIs in the cells in each metacell
umisumsdtsum <- umisumsdt[,.(total_umis=sum(total_umis_per_cell)),mc]
```

By combining the previous criteria we identify a pool of potentially bad metacells as those with non-zero UMI counts and total UMI counts below respective distribution peak. 

```{r mc_filt_bad}
# combine the metrics
dt <- csumsdtmean[umisumsdtsum,on="mc"]

# non-zero UMI genes threshold
dnonzeroumi <- density(dt$non_zero_genes_peak,adjust=0.5)
nonzerothrs <- 30

# total UMI threshold
dumi <- density(dt$total_umis,adjust=0.3)
totumithrs <- 12000
```

As a final criteria, we retain those potentially bad metacells that express more than 2 known TFs, if they also have moe than 2k UMIs

* **expression of known TFs**

```{r mc_filt_tf}
TFs_dt <- fread(file.path(annot_dir,"000TF_ANNOTATION_NEW"), header=FALSE)

# number of TFs expressed in metacells
tfs <- intersect(TFs_dt$V1,rownames(mc@mc_fp))
tfsmc <- apply(mc@mc_fp[tfs,],2,function(x) sum(x>2.5))
```

Inspect metacells' statistics. 

```{r}
tfdt <- data.table(mc=names(tfsmc),num_TFs=tfsmc)
dtt <- dt[tfdt,on="mc"]
# thresholding metacells
dtt[,status:="good"]
dtt[non_zero_genes_peak<nonzerothrs & total_umis<totumithrs, status:="bad"]
dtt[num_TFs>tfthrs & total_umis>2000, status:="good"]
dtt[,num_TFs_scaled:=pmin(num_TFs,5)]
gp <- ggplot(dtt,aes(non_zero_genes_peak,total_umis,label=mc,color=status)) + 
  geom_hline(yintercept=totumithrs, linetype="dashed") +  
  geom_vline(xintercept = nonzerothrs, linetype="dashed") +
  geom_point(aes(size=num_TFs_scaled),alpha=0.9) + scale_size(breaks=1:5,labels=c(1:4,">4"),name="# TFs") +
  geom_text_repel(show.legend=FALSE) + 
  scale_y_log10() + scale_x_log10() +
  labs(x="non-zero UMIs", y="total UMIs") +
  guides(colour=guide_legend(title="metacell",override.aes=list(size=6)))
```

Remove bad metacells.

```{r mc_kill}
kill_mcs <- dtt[status=="bad",mc]
writeLines(kill_mcs,file.path(scdb_dir,"kill_mcs.txt"))
mc <- mc_set_outlier_mc(mc,as.integer(kill_mcs))
mc <- mc_update_stats(mc,mat)
scdb_add_mc("mc_kill",mc)
```

# Cell type annotations

Looking at the expression of marker genes we assign cell types and colors to metacells - we save this in a tsv annotation file. We then reorder metacells and add cell type annotations to mc object, and also save final ordered annotation file.

```{r ann_reord}
mc_id <- "mc_kill"
mat_id <- "nvec_100"
mc <- scdb_mc(mc_id)
mat <- scdb_mat(mat_id)

annk <- fread(file.path(scdb_dir,"Metacell_annotation_reord"))
mc_reord <- mc_reorder(mc, as.integer(annk$metacell))
mc_reord@colors <- annk$color
scdb_add_mc("mc_nvec",mc_reord)
ann_final <- file.path(scdb_dir,"Nvec_metacell_annotation")
fwrite(annk[,metacell:=1:.N], ann_final, sep="\t")
```

# Save metacell output files

```{r mc_save}
mc_id <- "mc_nvec"
mat_id <- "nvec_100"
mc <- scdb_mc(mc_id)
mat <- scdb_mat(mat_id)

ann <- fread(file.path(scdb_dir, "Nvec_metacell_annotation"), sep="\t")

write.table(
  mc@mc_fp, file=file.path(scdb_dir, "Nvec_metacell_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

mc_umis_list <- lapply(colnames(mc@mc_fp), function(x) {
  cells <- names(mc@mc)[mc@mc==x]
  Matrix::rowSums(mat@mat[,cells])
})
mc_umis <- do.call(cbind,mc_umis_list)
colnames(mc_umis) <- colnames(mc@mc_fp)
mc_umifrac <- mc_umis/rep(colSums(mc_umis),each=nrow(mc_umis))

write.table(
  mc_umis, file=file.path(scdb_dir, "Nvec_metacell_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  mc_umifrac, file=file.path(scdb_dir, "Nvec_metacell_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc@cell_names[!(mc@cell_names %in% mc@outliers)],
  metacell = mc@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Nvec_metacell_assignments"), sep="\t")

```

# Metacells visualization

```{r mcel_vis_setup}
mc_id <- "mc_nvec"
mat_id <- "nvec_100"
mc <- scdb_mc(mc_id)
mat <- scdb_mat(mat_id)

bkg <- readLines(file.path(annot_dir, "black_list_genes"))
if (all(!grepl("$Nvec_",bkg)))
  bkg <- paste0("Nvec_",bkg)
  
ann <- fread(file.path(scdb_dir,"Metacell_annotation_081820"))
setnames(ann,c("metacell","cell_type","color"))
niche_order <- as.character(ann$metacell)
niche_color <- ann$color
# niche_order <- colnames(mc@mc_fp)
```

2D projection of metacells.

```{r 2d_mc_proj}
tgconfig::override_params(file.path(scdb_dir,"nvec_v3_metacell_params.yaml"),"metacell")
mcell_mc2d_force_knn(
  mc2d_id = sprintf("%s.2dproj",mc_id),
  mc_id = mc_id, 
  graph_id = "graphk100"
)
mcell_mc2d_plot(mc2d_id=sprintf("%s.2dproj",mc_id))
```

2D projections separated by input datasets.

```{r}
tgconfig::set_param(param = "mcell_mc2d_cex", value=1, package="metacell")
mcell_mc2d_plot_by_factor(
  mc2d_id = sprintf("%s.2dproj",mc_id),
  mat_id = mat_id,
  meta_field = "dataset"
)
```

Gene expression heatmaps. 

```{r}
hms <- scp_plot_cmod_markers(
  mc_object = mc, mat_object = mat,
  clust_ord = niche_order, clust_col = niche_color,
  black_list = c(grep("orphan",rownames(mc@mc_fp),value=TRUE),bkg),
  output_file=file.path(scdb_fig_dir,sprintf("Gene_expression_%s",mc_id)),
  width = 4200, height = 18000, plot_sc_width_cex = 5 ,plot_sc_height_cex = 1.5,
  show_gene_names = TRUE, gene_font_size = 18, gene_chr_limit = 50,
  mc_font_size = 20, mc_label_cex = 2,
  gene_annot_file=file.path(annot_dir,"Nvec_annotation_v2_2020-08-18"), 
  per_clust_genes=20, gene_min_fold=3, transverality_N = round(ncol(mc@mc_fp)/2),
  plot_sc=FALSE, print_heatmap=FALSE
)
```
```{r single_cell_expression_heatmap}
bl<-c(grep("orphan",rownames(mc@mc_fp),value=TRUE),bkg)
scr_plot_cmod_markers(
  mc_object = mc, mat_object = mat, 
  output_file=file.path(scdb_fig_dir,sprintf("Gene_expression_scr_%s",mc_id)),
  gene_annot_file=file.path(annot_dir,"Nvec_annotation_v2_2020-08-18"), 
  clust_ord=niche_order, clust_col=niche_color, include_peaks = FALSE,
  per_clust_genes=20, gene_min_fold=2, width=8000, black_list=bl, plot_sc=TRUE
)
```

Marker genes heatmaps.  

```{r mc_genes_heatmap}
scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, clust_ord = niche_order,
  markers_file = file.path(annot_dir, "000TF_ANNOTATION_NEW"),
  #markers_file = file.path(annot_dir, "all_tfs_2020-07-08_groups_Nvec_curated.csv"),
  heatmap_file = file.path(scdb_fig_dir, sprintf("Nvec_%s_TFs.png",mc_id)),
  marker_label_cex=3.5, mc_label_cex=4,
  chisq_threshold=1, T_totumi=10, w=5500, min_gene_fc=1.7, pmin=4
)

scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, 
  markers_file = file.path(annot_dir, "Nvec_ECM_prots_clean"), 
  heatmap_file = file.path(scdb_fig_dir,sprintf("Nvec_%s_ECM.png",mc_id)),
  marker_label_cex=3.5,mc_label_cex=4,
  chisq_threshold=1, T_totumi=10, w=5500, min_gene_fc=2, pmin=4
)

scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, 
  markers_file = file.path(annot_dir, "FINAL_neural_markers"),
  heatmap_file = file.path(scdb_fig_dir,sprintf("Nvec_%s_neuronal.png",mc_id)),
  marker_label_cex=3.5,mc_label_cex=4,
  chisq_threshold=1,T_totumi=10,w=5500,min_gene_fc=1.7,pmin=4
)
scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, 
  markers_file = file.path(annot_dir, "FINAL_synapse"),
  heatmap_file = file.path(scdb_fig_dir,sprintf("Nvec_%s_synapse.png",mc_id)),
  marker_label_cex=3,mc_label_cex=4,
  chisq_threshold=1,T_totumi=10,w=5500,min_gene_fc=1.7,pmin=4
)
scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, 
  markers_file = file.path(annot_dir, "Ciliogenesis_markers"),
  heatmap_file = file.path(scdb_fig_dir,sprintf("Nvec_%s_ciliogenesis.png",mc_id)),
  marker_label_cex=3.5,mc_label_cex=4,
  chisq_threshold=1,T_totumi=10,w=5500,min_gene_fc=1.7,pmin=4
)
scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, 
  markers_file = file.path(annot_dir, "FINAL_Cnidocytes"),
  heatmap_file = file.path(scdb_fig_dir,sprintf("Nvec_%s_cnido.png",mc_id)),
  marker_label_cex=3.5,mc_label_cex=4,
  chisq_threshold=1,T_totumi=10,w=5500,min_gene_fc=1.7,pmin=4
)
scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, 
  markers_file = file.path(annot_dir, "FINAL_muscle"),
  heatmap_file = file.path(scdb_fig_dir,sprintf("Nvec_%s_muscle.png",mc_id)),
  marker_label_cex=3,mc_label_cex=4,
  chisq_threshold=1,T_totumi=10,w=5500,min_gene_fc=1.7,pmin=4
)
```
```{r}
scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, clust_ord = niche_order,
  markers_file = file.path(annot_dir, "curated_GPCR_Nvec.csv"),
  heatmap_file = file.path(scdb_fig_dir, sprintf("Nvec_%s_GPCR.png",mc_id)),
  marker_label_cex=3.5, mc_label_cex=4,
  chisq_threshold=1, T_totumi=10, w=5500, min_gene_fc=2, pmin=4
)
scr_barplot_heatmap_markers(
  mc_object = mc, mat_object = mat, clust_ord = niche_order,
  markers_file = file.path(annot_dir, "curated_ionchannels_Nvec_SD1.csv"),
  heatmap_file = file.path(scdb_fig_dir, sprintf("Nvec_%s_ionchannel.png",mc_id)),
  marker_label_cex=3.5, mc_label_cex=4,
  chisq_threshold=1, T_totumi=10, w=5500, min_gene_fc=2, pmin=4
)
```

Expression of genes in metacells

```{r}
neuro_genes_1 <- c(
  "Pou4"="v1g160868","FoxL2"="v1g67043","ZicC zfC2H2"="v1g121215",
  "Shak1"="v1g21645","SHak4"="v1g189854","Nav2.5"="NVE7349"
)
neuro_genes_2 <- c(
  "OtxC"="v1g213735","SoxC"="v1g131609","Ets"="v1g16815","Islet"="v1g161808","GATA"="v1g196609"
)
neuro_genes_3 <- c("Delilah"="v1g39308","Vasopresin"="v1g39308","Shal"="v1g1485","GPCR AMPA"="v1g104623")
neuro_genes_4 <- c("Nk6"="v1g139174","GABA"="v1g33294","Unclass Sox" ="v1g110350","zfC2H2"="v1g66221")
gi <- c(neuro_genes_1,neuro_genes_2,neuro_genes_3,neuro_genes_4)
gin <- paste0("Nvec_",gi)
names(gin) <- names(gi)
pdf(file.path(scdb_fig_dir,"neuro_genes_expression.pdf"),width=12,height=5+length(gi)*3,useDingbats=TRUE)
par(mfrow=c(length(gin),1))
for (i in 1:length(gin)) {
print(barplot(
  mc@mc_fp[gin[i], colnames(mc@mc_fp) %in% ann[grepl("Neuro",cell_type),metacell]],
  col = ann[grepl("Neuro",cell_type),color],
  las=2, main=paste(names(gi)[i], gi[i])
))
}
dev.off()
```

# Cell type level metacell analysis

```{r ct_mc}
mc_id <- "mc_nvec"
mat_id <- "nvec_100"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
ann_file <- file.path(scdb_dir,"Nvec_metacell_annotation")
ann <- fread(ann_file)

# calculate mc object with mc_fp by cell type
mc_ct <- sca_cell_type_fp(
  input_table = ann_file,
  mc_object = mc,
  mat_object = mat
)
scdb_add_mc(id="ct_nvec",mc_ct)

# save footprint
write.table(
  mc_ct@mc_fp, file=file.path(scdb_dir, "Nvec_cell_type_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

# counts
ct_umis_list <- lapply(colnames(mc_ct@mc_fp), function(y) {
  x <- ann[cell_type==y,]$metacell
  cells <- names(mc_ct@mc)[mc_ct@mc%in%x]
  Matrix::rowSums(mat@mat[,cells])
})
ct_umis <- do.call(cbind,ct_umis_list)
colnames(ct_umis) <- colnames(mc_ct@mc_fp)
ct_umifrac <- ct_umis/rep(colSums(ct_umis),each=nrow(ct_umis))

write.table(
  ct_umis, file=file.path(scdb_dir, "Nvec_cell_type_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  ct_umifrac, file=file.path(scdb_dir, "Nvec_cell_type_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc_ct@cell_names[!(mc_ct@cell_names %in% mc@outliers)],
  cell_type = mc_ct@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Nvec_cell_type_assignments"), sep="\t")

```

# Broad cell type level metacell analysis

```{r ct_mc}
#source("scr_plot_cmod_markers.R")
mc_id <- "ct_081820"
mat_id <- "nvec_100"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
ann_file <- file.path(scdb_dir,"Nvec_broad_cell_type_annotation")
ann <- fread(ann_file)

# calculate mc object with mc_fp by cell type
mc_ct <- sca_cell_type_fp(
  input_table = ann_file,
  mc_object = mc,
  mat_object = mat
)
scdb_add_mc(id="bct_nvec",mc_ct)

# save footprint
write.table(
  mc_ct@mc_fp, file=file.path(scdb_dir, "Nvec_broad_cell_type_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

# counts
ct_umis_list <- lapply(colnames(mc_ct@mc_fp), function(y) {
  x <- ann[cell_type==y,]$metacell
  cells <- names(mc_ct@mc)[mc_ct@mc%in%x]
  Matrix::rowSums(mat@mat[,cells])
})
ct_umis <- do.call(cbind,ct_umis_list)
colnames(ct_umis) <- colnames(mc_ct@mc_fp)
ct_umifrac <- ct_umis/rep(colSums(ct_umis),each=nrow(ct_umis))

write.table(
  ct_umis, file=file.path(scdb_dir, "Nvec_broad_cell_type_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  ct_umifrac, file=file.path(scdb_dir, "Nvec_broad_cell_type_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc_ct@cell_names[!(mc_ct@cell_names %in% mc@outliers)],
  cell_type = mc_ct@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Nvec_broad_cell_type_assignments"), sep="\t")

```

# Sequencing and clustering statistics

```{r}
cells_final <- fread(file.path(scdb_dir,"Nvec_broad_cell_type_assignments"))[[1]]
wells_cells <- fread("/users/asebe/asebe/proj/scRNA_nvec_v3/scdb_v3/annotations/wells_cells.txt")
batch_table <- fread("/users/asebe/asebe/proj/scRNA_nvec_v3/scdb_v3/annotations/amp_batches.txt")
files=list.files("/users/asebe/asebe/proj/scRNA_nvec_v3/scdb_v3/output_good/QC/read_stats/",pattern="*.txt",full.names=T)
umi_files=list.files(path="/users/asebe/asebe/proj/scRNA_nvec_v3/scdb_v3/output_good/umi.tab/",pattern="*.txt",full.names=T)
cell_Tumis_list <- lapply(umi_files, function(f) { 
  ff=fread(f, header=TRUE)[,-1:2]
  colSums(as.matrix(ff))
})
cell_Tumis <- do.call('cbind',cell_Tumis_list)
cell_Tumis <- as.data.table(cell_Tumis,keep.rownames="well_id")[match(cells_final,well_id)]
# mapping stats per cell
cell_stats <- rbindlist(lapply(files, fread))
cell_stats <- cell_stats[well_id %in% wells_cells$Well_ID]
cols <- c("mapped_to_nongenic", grep("gene",colnames(cell_Treads),value=TRUE))
other_mapped <- rowSums(cell_stats[,..cols])
cell_stats[,total_mapped:=cell_stats$gene_mapped+other_mapped]
fcols <- c("well_id", "total","total_mapped","gene_mapped")
cell_Treads <- cell_stats[match(cells_final,well_id),..fcols]
dt <- merge.data.table(cell_Tumis,cell_Treads,by="well_id")
dt[,cells_reads_umi:=total_mapped/cell_Tumis]
setnames(dt,
  c("well_id","cell_Tumis","total","total_mapped","cells_reads_umi"),
  c("cell","UMI_per_cell","total_reads_per_cell","mapped_reads_per_cell","mapped_reads_per_UMI")
)
fwrite(dt,file.path(scdb_dir,"stats_cells.tsv"),sep='\t')
dtmed <- dt[,lapply(.SD, median),.SDcols=colnames(dt)[-1]]
dtmed[,':='(dataset="Nvec_v3",cells=nrow(dt))]
setcolorder(dtmed,c("dataset","cells"))
fwrite(dtmed,file.path(scdb_dir,"stats.tsv"),sep='\t')
```

# Session info

```{r session, eval=TRUE}
sessionInfo()
```
