---
title: "06. Xenia sp. metacell analysis"
date: "`r format(Sys.time(), '%d %B, %Y')`"
auhor: "Anamaria Elek"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: false
    toc_depth: 2
---
```{r options, include=FALSE}
knitr::opts_chunk$set(eval = FALSE, fig.align = 'center')
```

# Set up metacell analysis

Load required packages and functions.

```{r setup, eval=TRUE, message=FALSE, warning=FALSE}
library(data.table)
library(ggplot2)
library(ggrepel)
library(ggrastr)
library(plotly)
library(scales)
library(stringr)
library(RColorBrewer)
library(circlize)
library(tgconfig)
library(metacell)
theme_py <- theme_classic() + theme(
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA),
  text = element_text(size=20)
)
theme_set(theme_py)
# load functions for downstream analysis
scripts <- list.files("metacell_downstream_functions",full.names=TRUE)
for (file in scripts) source(file)
```

Initialize database and figures output directory.

```{r mc_setup, eval=TRUE, warning=FALSE, message=FALSE}
require(metacell)

# analysis dir
cluster_dir <- "clustering_xenia"
dir.create(cluster_dir, showWarnings=FALSE)

# initialize db directory
scdb_dir <- file.path(cluster_dir, "scdb")
dir.create(scdb_dir, showWarnings=FALSE)
scdb_init(scdb_dir, force_reinit=TRUE)

# plotting directory
scdb_fig_dir <- file.path(scdb_dir,"figs")
dir.create(scdb_fig_dir, showWarnings=FALSE)
scfigs_init(scdb_fig_dir)

# directory with annotations
annot_dir <- "annotation_xenia"
```

# Metacell pipeline
## Input data filtering

Load the expression data to metacell database.  

```{r mcell_import}
# import UMI table
mcell_import_multi_scmat_10x(mat_nm="mat_nowholev2",dataset_table_fn="10x_import_table.txt",force=T,base_dir="/users/asebe/asebe/proj/scRNA_xenia/Cellranger/")
mat_nowholev2=scdb_mat("mat_nowholev2")
kill_nowholev2=setdiff(colnames(mat_nowholev2@mat),orig_cell_names)
mcell_mat_ignore_cells("mat_nowholev2","mat_nowholev2",ig_cells=kill_nowholev2)
mat_nowholev2=scdb_mat("mat_nowholev2")
```

Filter by cell depth (UMIs/cell)

```{r mcell_filter}
erccs=rownames(mat_nowholev2@mat)[grepl("ERCC",rownames(mat_nowholev2@mat))]
peak_counts=rowSums(as.matrix(mat_nowholev2@mat[grepl("peak_",rownames(mat_nowholev2@mat)),]))
peaks_kill=names(which(peak_counts < 300))
mcell_mat_ignore_genes("mat_nowholev2","mat_nowholev2",ig_genes=c(peaks_kill,erccs))
mat_nowholev2=scdb_mat("mat_nowholev2")
```

Calculate gene statistics

```{r mcell_gstat}
mcell_add_gene_stat(gstat_id = "gstat", mat_id = mat_nm_filt, force = TRUE)
mat <- scdb_mat(mat_nm_filt)
```

## Building metacells

Select marker genes for building metacell graph and Build metacells.

```{r mcell_markers}
mcell_add_gene_stat(gstat_id="gstat_nowholev2", mat_id="mat_nowholev2", force=T)
mcell_gset_filter_multi("gstat_nowholev2","clust_markers_nowholev2",T_tot=30,T_top3=2,T_szcor=-0.08,T_niche=0.01,force_new=T)
mcell_add_cgraph_from_mat_bknn(mat_id="mat_nowholev2",gset_id="clust_markers_nowholev2",graph_id="graphk100_nowholev2",K=100,dsamp=F)
mcell_coclust_from_graph_resamp(coc_id="coc500_min20_nowholev2",graph_id="graphk100_nowholev2",min_mc_size=20,p_resamp=0.75,n_resamp=100)
mcell_mc_from_coclust_balanced("coc500_min20_nowholev2",mat="mat_nowholev2",mc_id="mc_k30_nowholev2",K=30,min_mc_size=20,alpha=2)

```

Batch distribution in v3+v2_nowhole

```{r}
mc_nowholev2=scdb_mc("mc_k30_nowholev2")
mat_nowholev2=scdb_mat("mat_nowholev2")
mat_nowholev2@cell_metadata=mat_nowholev2@cell_metadata[colnames(mat_nowholev2@mat),]
mat_nowholev2@cell_metadata$color=as.character(rep(c("darkblue","darkred","darkorange"),table(factor(mat_nowholev2@cell_metadata$batch_set_id,levels=unique(mat_nowholev2@cell_metadata$batch_set_id)))))
scr_batch_distribution(mc_nowholev2,mat_nowholev2,1:ncol(mc_nowholev2@mc_fp),batch_field="batch_set_id")

gset=scdb_gset("clust_markers_nowholev2")
gset=names(gset@gene_set)
scr_plot_cmod_markers(mc_nowholev2,mat_nowholev2,clust_ord=1:ncol(mc_nowholev2@mc_fp),gene_list=gset,output_file="Global_gene_expression_v3only_feature_genes.png",gene_annot_file="Annot_gene_lists/Xesp_annotation_v2_2020-06-25.csv",per_clust_genes=50,height=15000,width=6000,gene_min_fold=1.5,plot_sc=T,add_lateral_genes=F)
```

Identify strongly batch-related genes.

```{r}
gset=scdb_gset("clust_markers_nowholev2")
gset=names(gset@gene_set)

#we first calculate the frequency of each batch in each metacell
mc_nowholev2=scdb_mc("mc_k30_nowholev2")
mat_nowholev2=scdb_mat("mat_nowholev2")
m=matrix(nrow=3,ncol=ncol(mc_nowholev2@mc_fp),unlist(tapply(mat_nowholev2@cell_metadata[names(mc@mc),"batch_set_id"],mc_nowholev2@mc,function(x) table(factor(x,levels=unique(mat_nowholev2@cell_metadata$batch_set_id))))))
m_f=t(apply(m,1,function(x) x/colSums(m)))
rownames(m_f)=unique(mat_nowholev2@cell_metadata$batch_set_id)

g_cor_whole3=cor(t(mc_nowholev2@mc_fp),m_f["whole_v3",],method="spearman")

#the main batch effect is observed in v3 vs v2, hence we'll pick only genes with whole_v3 association:
batchy=union(names(which(g_cor_whole3[,1]>quantile(g_cor_whole3,0.97))),names(which(g_cor_whole3[,1]<quantile(g_cor_whole3,0.03))))
write.table(batchy,file.path(annotation_dir,"xenia_bl"),quote=F,col.names=F,row.names=F)
```

Clustering v3+v2_no_whole

```{r}
batchy=scan("tmp_bl_3perc",what="")
mcell_gset_filter_multi("gstat_nowholev2","clust_markers_nowholev2_bl",T_tot=30,T_top3=2,T_szcor=-0.08,T_niche=0.01,force_new=T,bl=batchy)
mcell_add_cgraph_from_mat_bknn(mat_id="mat_nowholev2",gset_id="clust_markers_nowholev2_bl",graph_id="graphk100_nowholev2_bl",K=100,dsamp=F)
mcell_coclust_from_graph_resamp(coc_id="coc500_min20_nowholev2_bl",graph_id="graphk100_nowholev2_bl",min_mc_size=20,p_resamp=0.75,n_resamp=100)
mcell_mc_from_coclust_balanced("coc500_min20_nowholev2_bl",mat="mat_nowholev2",mc_id="mc_k30_nowholev2_bl",K=30,min_mc_size=20,alpha=2)
```

Check batch distribution in v3+v2_nowhole after removing blacklisted genes.

```{r}
mc_nowholev2=scdb_mc("mc_k30_nowholev2_bl")
mat_nowholev2=scdb_mat("mat_nowholev2")
mat_nowholev2@cell_metadata=mat_nowholev2@cell_metadata[colnames(mat_nowholev2@mat),]
mat_nowholev2@cell_metadata$color=as.character(rep(c("darkblue","darkred","darkorange"),table(factor(mat_nowholev2@cell_metadata$batch_set_id,levels=unique(mat_nowholev2@cell_metadata$batch_set_id)))))
scr_batch_distribution(mc_nowholev2,mat_nowholev2,1:ncol(mc_nowholev2@mc_fp),batch_field="batch_set_id")

scr_plot_cmod_markers(mc_nowholev2,mat_nowholev2,clust_ord=1:ncol(mc_nowholev2@mc_fp),gene_list=gset,output_file="Global_gene_expression_v3only_feature_genes.png",gene_annot_file="Annot_gene_lists/Xesp_annotation_v2_2020-06-25.csv",per_clust_genes=50,height=15000,width=6000,gene_min_fold=1.5,plot_sc=T,add_lateral_genes=F)
```

## Filtering bad metacells

```{r mcell_load_filt}
mc_id <- "mc_k30_nowholev2_bl"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
mcs <- colnames(mc@mc_fp)
marker_genes_gset <- scdb_gset(id="clust_markers_nowholev2_bl")
marker_genes <- names(marker_genes_gset@gene_set)
```

The following criteria are used for filtering out bad metacells:

* **number of genes with non-zero UMIs** summarized across all cells in metacell - metacells consisting of cells that express very low number of genes are potentially bad metacells

```{r nonzero_umi}
# count genes with non-zero UMIs in single cells
csums <- lapply(mcs, function(mcid) {
  mc_cells <- names(mc@mc[mc@mc==mcid])
  mc_cells <- intersect(mc_cells,colnames(mat@mat))
  nonzerogenes <- Matrix::colSums(mat@mat[marker_genes,mc_cells]>0)
  data.table(non_zero_genes_per_cell = nonzerogenes)
})
csumsdt <- rbindlist(csums,idcol = "mc")
csumsdt[,mc:=factor(mc,levels=mcs)]
# summarize by taking the peak of distribution of non-zero UMI counts in the cells in each metacell
csumsdtmean <- csumsdt[,.SD[,.(
  non_zero_genes_peak={
    dns = density(non_zero_genes_per_cell)
    ind = which.max(dns$y)
    dns$x[ind]
})],mc]
```

* **total number of UMIs** summed across all cells in metacell - metacells consisting of very small cells, with a low UMI count, are potentially bed metacells.

```{r total_umi}
# count total UMIs in cells
umisums <- lapply(mcs, function(mcid) {
  mc_cells <- names(mc@mc[mc@mc==mcid])
  mc_cells <- intersect(mc_cells,colnames(mat@mat))
  mc_table <- mat@mat[marker_genes,mc_cells]
  totalumis <- Matrix::colSums(mc_table)
  data.table(total_umis_per_cell = totalumis)
})
umisumsdt <- rbindlist(umisums,idcol = "mc")
umisumsdt[,mc:=factor(mc,levels=mcs)]
# summarize by taking the sum of total UMIs in the cells in each metacell
umisumsdtsum <- umisumsdt[,.(total_umis=sum(total_umis_per_cell)),mc]
```

By combining the previous criteria we identify a pool of potentially bad metacells as those with non-zero UMI counts and total UMI counts below respective distribution peak. 

```{r mc_filt_bad}
# combine the metrics
dt <- csumsdtmean[umisumsdtsum,on="mc"]

# non-zero UMI genes threshold
dnonzeroumi <- density(dt$non_zero_genes_peak,adjust=0.4)
nonzerothrs <- 70

# total UMI threshold
dumi <- density(dt$total_umis,adjust=0.3)
totumithrs <- 16000
```

As a final criteria, we retain those potentially bad metacells that express more than 2 known TFs, if they also have moe than 1k UMIs

* **expression of known TFs**

```{r mc_filt_tf}
TFs_dt <- fread(file.path(annot_dir,"all_tfs_2020-07-08_groups_Xesp_curated.csv"), header=FALSE)
# number of TFs expressed in metacells
tfs <- intersect(TFs_dt$V1,rownames(mc@mc_fp))
tfsmc <- apply(mc@mc_fp[tfs,],2,function(x) sum(x>2.5))
```

Inspect metacells' statistics. 

```{r}
tfdt <- data.table(mc=names(tfsmc),num_TFs=tfsmc)
dtt <- dt[tfdt,on="mc"]
# thresholding metacells
dtt[,status:="good"]
dtt[non_zero_genes_peak<nonzerothrs & total_umis<totumithrs, status:="bad"]
dtt[num_TFs>tfthrs & total_umis>1000, status:="good"]
dtt[,num_TFs_scaled:=pmin(num_TFs,5)]
gp <- ggplot(dtt,aes(non_zero_genes_peak,total_umis,label=mc,color=status)) + 
  geom_hline(yintercept=totumithrs, linetype="dashed") +  
  geom_vline(xintercept = nonzerothrs, linetype="dashed") +
  geom_point(aes(size=num_TFs_scaled),alpha=0.9) + scale_size(breaks=1:5,labels=c(1:4,">4"),name="# TFs") +
  geom_text_repel(show.legend=FALSE) + 
  scale_y_log10() + scale_x_log10() +
  labs(x="non-zero UMIs", y="total UMIs") +
  guides(colour=guide_legend(title="metacell",override.aes=list(size=6)))
```

Remove bad metacells.

```{r mc_kill}
kill_mcs <- dtt[status=="bad",mc]
writeLines(kill_mcs,file.path(scdb_dir,"kill_mcs.txt"))
mc <- mc_set_outlier_mc(mc,as.integer(kill_mcs))
mc <- mc_update_stats(mc,mat)
scdb_add_mc("mc_kill",mc)
```

# Annotation correlations

Load data 

```{r load_data_comp}
mc_id <- "mc_kill"
mat_id <- "mat_nowholev2"
mc <- scdb_mc(mc_id)
mat <- scdb_mat(mat_id)
```

Nvec data

```{r load_nvec_comp, eval=TRUE}
old_mcfp <- read.table("clustering_nematostella/Nvec_metacell_gene_FC")
old_assign <- fread("clustering_nematostella/Nvec_metacell_assignments")
old_annot <- "clustering_nematostella/Nvec_metacell_annotation"
setnames(old_assign,c("cell","metacell"))
colnames(old_mcfp) <- str_remove(colnames(old_mcfp),"X")
var_genes <- names(which(apply(old_mcfp,1,max)>2))
```

Correlation with Nvec

```{r nvec_comp}
cor_method <- "spearman"
comp_method = "jaccard"
cors <- sca_clustering_correlation(
  matrix1 = old_mcfp, matrix2 = mc@mc_fp, 
  markers = var_genes, 
  original_ordering_1st = TRUE, original_ordering_2nd = TRUE,
  annotation_file_1 = old_annot,
  cor_method = cor_method, pmin = 0.8,
  output_file = file.path(scdb_fig_dir,sprintf("xenia_%s_vs_nvec_%s.png",mc_id,cor_method)),
  height = 1000, width = 1200
)
lfc_threshold <- 1.2
jacc <- sca_clustering_comparison(
  matrix1 = old_mcfp, matrix2 = mc@mc_fp, markers = var_genes, lfc_threshold = lfc_threshold,
  original_ordering_1st = TRUE, original_ordering_2nd = TRUE, comp_method = comp_method, 
  annotation_file_1 = old_annot,
  output_file = file.path(
    scdb_fig_dir,sprintf("xenia_%s_vs_nvec_%s_%s.png",mc_id,comp_method,lfc_threshold)
  ), height = 2300, width = 2400
)
```

Save cell and metacell annotations.  

```{r annot_files_save}
mc <- scdb_mc("mc_filt")
mc_annotation <- unique(cell_annotation[,.(mc,cell_type,colour)])
mc_cell_info <- cell_annotation[,.(cell,mc,cell_type,colour)]

mc_annotation_file <- file.path(scdb_dir,"Xenia_metacell_annotation")
mc_cell_info_file <- file.path(scdb_dir,"Xenia_cell_annotoation")

fwrite(mc_annotation, mc_annotation_file, sep="\t")
fwrite(mc_cell_info, mc_cell_info_file, sep="\t")
```

Modify transferrerd annotations after inspecting 2d plot and heatmap and update metacell object.

```{r fix_cols}
mc_annotation_file <- file.path(scdb_dir,"Xenia_metacell_annotation")
mc_annotation <- fread(mc_annotation_file, sep="\t")

# update metacells colors
colid <- match(mc@annots,mc_annotation$mc)
niche_col_new <- mc_annotation[colid,colour]
mc@colors <- niche_col_new

# add new mc object to db
scdb_add_mc(id="mc_filt", mc=mc)
```

# Cell type annotations

Load objects from metacell database

```{r transfer_annots_setup}
mc_id <- "mc_filt"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
```

Create reordered, filtered and annotated metacell object.

```{r mc_reord_annot}
annk <- fread("Xesp_metacell_annotation_reord")

# reorder
mc_reord <- mc_reorder(mc, as.integer(annk$metacell))
mc_reord@colors <- annk$color

# kill
mc_kill <- mc_set_outlier_mc(mc_reord,which(annk$cell_type=="KILL"))
mc_kill <- mc_update_stats(mc_kill,mat)
ann <- annk[cell_type!="KILL"]

# add mc object
scdb_add_mc("mc_xenia",mc_kill)

# save final annots
ann_final <- file.path(scdb_dir,"Xesp_metacell_annotation")
fwrite(annk[,cell_type!="KILL"][,metacell:=1:.N], ann_final, sep="\t")
```

Save metacell output files.

```{r mc_save}
mc_id <- "mc_xenia"
mat_id <- "mat_nowholev2"
mc <- scdb_mc(mc_id)
mat <- scdb_mat(mat_id)
ann_file <- file.path(scdb_dir,"Xesp_metacell_annotation")
ann <- fread(ann_file)

write.table(
  mc@mc_fp, file=file.path(scdb_dir, "Xesp_metacell_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

mc_umis_list <- lapply(colnames(mc@mc_fp), function(x) {
  cells <- names(mc@mc)[mc@mc==x]
  Matrix::rowSums(mat@mat[,cells])
})
mc_umis <- do.call(cbind,mc_umis_list)
colnames(mc_umis) <- colnames(mc@mc_fp)
mc_umifrac <- mc_umis/rep(colSums(mc_umis),each=nrow(mc_umis))

write.table(
  mc_umis, file=file.path(scdb_dir, "Xesp_metacell_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  mc_umifrac, file=file.path(scdb_dir, "Xesp_metacell_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc@cell_names[!(mc@cell_names %in% mc@outliers)],
  metacell = mc@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Xesp_metacell_assignments"), sep="\t")

```


# Cell type level metacell analysis

Summarize metacells across cell types, and add to the database.

```{r ct_mc}
mc_id <- "mc_xenia"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
ann_file <- file.path(scdb_dir,"Xesp_metacell_annotation")
ann <- fread(ann_file)

# calculate mc object with mc_fp by cell type
mc_ct <- sca_cell_type_fp(
  input_table=ann_file,
  mc_object = mc,
  mat_object = mat
)
scdb_add_mc(id="ct_xenia",mc_ct)

# save footprint
write.table(
  mc_ct@mc_fp, file=file.path(scdb_dir, "Xesp_cell_type_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

# counts
ct_umis_list <- lapply(colnames(mc_ct@mc_fp), function(y) {
  x <- ann[cell_type==y,]$metacell
  cells <- names(mc_ct@mc)[mc_ct@mc%in%x]
  Matrix::rowSums(mat@mat[,cells])
})
ct_umis <- do.call(cbind,ct_umis_list)
colnames(ct_umis) <- colnames(mc_ct@mc_fp)
ct_umifrac <- ct_umis/rep(colSums(ct_umis),each=nrow(ct_umis))

write.table(
  ct_umis, file=file.path(scdb_dir, "Xesp_cell_type_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  ct_umifrac, file=file.path(scdb_dir, "Xesp_cell_type_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc_ct@cell_names[!(mc_ct@cell_names %in% mc@outliers)],
  cell_type = mc_ct@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Xesp_cell_type_assignments"), sep="\t")

```

# Broad cell type level metacell analysis

Group  cell types into broad cell types, and add to the database.

```{r bct_mc}
mc_id <- "ct_xenia"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
ann_file <- file.path(scdb_dir,"Xesp_broad_cell_type_annotation")
ann <- fread(ann_file)

# calculate mc object with mc_fp by cell type
mc_ct <- sca_cell_type_fp(
  input_table = ann_file,
  mc_object = mc,
  mat_object = mat
)
scdb_add_mc(id="bct_xenia",mc_ct)

# save footprint
write.table(
  mc_ct@mc_fp, file=file.path(scdb_dir, "Xesp_broad_cell_type_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

# counts
ct_umis_list <- lapply(colnames(mc_ct@mc_fp), function(y) {
  x <- ann[cell_type==y,]$metacell
  cells <- names(mc_ct@mc)[mc_ct@mc%in%x]
  Matrix::rowSums(mat@mat[,cells])
})
ct_umis <- do.call(cbind,ct_umis_list)
colnames(ct_umis) <- colnames(mc_ct@mc_fp)
ct_umifrac <- ct_umis/rep(colSums(ct_umis),each=nrow(ct_umis))

write.table(
  ct_umis, file=file.path(scdb_dir, "Xesp_broad_cell_type_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  ct_umifrac, file=file.path(scdb_dir, "Xesp_broad_cell_type_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc_ct@cell_names[!(mc_ct@cell_names %in% mc@outliers)],
  cell_type = mc_ct@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Xesp_broad_cell_type_assignments"), sep="\t")

```

# Sequencing and clustering statistics

Load objects from metacell database

```{r}
mc_id <- "mc_xenia"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)

bl <- readLines("annotation_xenia/xenia_bl")

ann <- fread(file.path(scdb_dir,"Xesp_metacell_annotation"))
setnames(ann,c("metacell","cell_type","color"))
niche_order <- as.character(ann$metacell)
niche_color <- ann$color
```

Extract stats for metacell clustering.

```{r}
# cell sizes
cell_sizes <- apply(mat@mat,2,sum)
cdt <- data.table(cells=colnames(mat@mat),cell_sizes=cell_sizes)
cdt[data.table(cells=names(mc@mc),metacells=mc@mc), on="cells", mc:=i.metacells]
cdt <- cdt[!is.na(mc)]
cdt[,median_mc_cell_size:=median(cell_sizes),mc]
cdt[,mc:=factor(mc,levels=sort(as.integer(unique(cdt$mc))))]
cdt[,total_mc_umis:=sum(cell_sizes),mc]
cdt[,total_mc_cells:=.N,mc]
fwrite(cdt,file.path(scdb_dir,"cdt.tsv"),sep="\t")
# gene sizes
gene_umis <- apply(mat@mat,1,sum)
gdt <- data.table(gene=rownames(mat@mat),gene_umis=gene_umis)
fwrite(gdt,file.path(scdb_dir,"gdt.tsv"),sep="\t")
# gene stats
gstat_id <- "gstat_nowholev2"
gstat <- scdb_gstat(gstat_id)
gset_id <- "clust_markers_nowholev2_bl"
marker_genes_gset <- scdb_gset(gset_id)
markers <- names(marker_genes_gset@gene_set)
gstat <- setDT(gstat)
setnames(gstat,"name","gene")
gstat[, marker:="no"]
gstat[gene %in% markers, marker:="yes"]
gstat[,log2_ds_mean:=log2(ds_mean)]
fwrite(gstat,file.path(scdb_dir,"gstat.tsv"),sep="\t")
# markers in cells
markers_mat <- mat@mat[markers,cdt$cells]
markers_num <- apply(markers_mat, 2, function(x) sum(x>0))
mdf <- cbind.data.frame(num_markers=markers_num)
fwrite(mdf,file.path(scdb_dir,"mdf.tsv"),sep="\t")
```

Count reads from CellRanger output bam files.

```{bash}
#!/bin/bash
#$ -V
#$ -cwd
#$ -N bamcount
#$ -M anamaria.elek@crg.eu
#$ -m a
#$ -o read_counts.out
#$ -j y
#$ -q long-sl7,mem_512_12h
#$ -l virtual_free=20G,h_rt=43200
#$ -pe smp 12

cores=4
ml purge
ml BEDTools/2.27.1-foss-2016b

BAMS=( ls *bam )
COUNTS="/clustering_xenia/scdb/read_counts.csv"
echo file,total_reads,mapped_reads > $COUNTS
for bam in $BAMS
do
  echo $bam
  countsc=${bam%%bam}.CB.counts.txt
  countscmq=${bam%%bam}mapq30.CB.counts.txt
  echo "Counting BCs"
  samtools view $bam | grep CB:Z: | sed 's/.*CB:Z:\([ACGT]*\).*/\1/' | sort | uniq -c > $countsc
  echo "Counting BCs for mapq30"
  samtools view -c -q 30 -F 260 $bam | grep CB:Z: | sed 's/.*CB:Z:\([ACGT]*\).*/\1/' | sort | uniq -c > $countscmq
  # TOTAL_READS=$( samtools view -c $bam )
  # MAPPED_READS=$( samtools view -c -q 30 -F 260 $bam )
  # echo ${bam},${TOTAL_READS},${MAPPED_READS} >> $COUNTS
done

```

Summarize

```{r}
cdt <- fread(file.path(scdb_dir,"cdt.tsv"))
cdt[,dataset:=str_extract(cells,'(tentacle|whole|stalk)_v\\d+')]
dt <- merge.data.table(
  cdt[,.(cells=.N),dataset],
  cdt[,.(UMI_per_cell=median(cell_sizes)),dataset],
  by="dataset"
)

counts <- rbindlist(lapply(list.files(scdb_dir,"read.counts.csv",full.names=TRUE)[-1],fread))
counts[,dataset:=str_extract(file,'(tentacle|whole|stalk)_v\\d+')][,file:=NULL]
stats <- merge.data.table(dt,counts,by="dataset",all.x=TRUE)
stats[,total_reads_per_cell:=total_reads/cells]
stats[,mapped_reads_per_cell:=mapped_reads/cells]
stats[,mapped_reads_per_UMI:=mapped_reads_per_cell/UMI_per_cell]
fwrite(stats,file.path(scdb_dir,"stats.tsv"),sep="\t")

```

# Session info

```{r session, eval=TRUE}
sessionInfo()
```

