---
title: "06. Xenia sp. metacell analysis"
date: "`r format(Sys.time(), '%d %B, %Y')`"
auhor: "Anamaria Elek"
output: 
  html_document:
    toc: true
    toc_float:
      collapsed: true
      smooth_scroll: false
    toc_depth: 2
---
```{r options, include=FALSE}
knitr::opts_chunk$set(eval = FALSE, fig.align = 'center')
```

# Set up metacell analysis

Load required packages and functions.

```{r setup, eval=TRUE, message=FALSE, warning=FALSE}
library(data.table)
library(ggplot2)
library(ggrepel)
library(ggrastr)
library(plotly)
library(scales)
library(stringr)
library(RColorBrewer)
library(circlize)
library(tgconfig)
library(metacell)
theme_py <- theme_classic() + theme(
  panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(),
  panel.border = element_rect(colour = "black", fill = NA),
  text = element_text(size=20)
)
theme_set(theme_py)
# load functions for downstream analysis
scripts <- list.files("metacell_downstream_functions",full.names=TRUE)
for (file in scripts) source(file)
```

Initialize database and figures output directory.

```{r mc_setup, eval=TRUE, warning=FALSE, message=FALSE}
require(metacell)

# analysis dir
cluster_dir <- "clustering_xenia"
dir.create(cluster_dir, showWarnings=FALSE)

# initialize db directory
scdb_dir <- file.path(cluster_dir, "scdb")
dir.create(scdb_dir, showWarnings=FALSE)
scdb_init(scdb_dir, force_reinit=TRUE)

# plotting directory
scdb_fig_dir <- file.path(scdb_dir,"figs")
dir.create(scdb_fig_dir, showWarnings=FALSE)
scfigs_init(scdb_fig_dir)

# directory with annotations
annot_dir <- "annotation_xenia"
```

# Metacell pipeline

## Filtering bad metacells

```{r mcell_load_filt, echo=FALSE, eval=TRUE}
mc_id <- "mc_k30_nowholev2_bl"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
mcs <- colnames(mc@mc_fp)
marker_genes_gset <- scdb_gset(id="clust_markers_nowholev2_bl")
marker_genes <- names(marker_genes_gset@gene_set)
```

The following criteria are used for filtering out bad metacells:

* **number of genes with non-zero UMIs** summarized across all cells in metacell - metacells consisting of cells that express very low number of genes are potentially bad metacells

```{r nonzero_umi, eval=TRUE}
# count genes with non-zero UMIs in single cells
csums <- lapply(mcs, function(mcid) {
  mc_cells <- names(mc@mc[mc@mc==mcid])
  mc_cells <- intersect(mc_cells,colnames(mat@mat))
  nonzerogenes <- Matrix::colSums(mat@mat[marker_genes,mc_cells]>0)
  data.table(non_zero_genes_per_cell = nonzerogenes)
})
csumsdt <- rbindlist(csums,idcol = "mc")
csumsdt[,mc:=factor(mc,levels=mcs)]
# summarize by taking the peak of distribution of non-zero UMI counts in the cells in each metacell
csumsdtmean <- csumsdt[,.SD[,.(
  non_zero_genes_peak={
    dns = density(non_zero_genes_per_cell)
    ind = which.max(dns$y)
    dns$x[ind]
})],mc]
```

* **total number of UMIs** summed across all cells in metacell - metacells consisting of very small cells, with a low UMI count, are potentially bed metacells.

```{r total_umi, eval=TRUE}
# count total UMIs in cells
umisums <- lapply(mcs, function(mcid) {
  mc_cells <- names(mc@mc[mc@mc==mcid])
  mc_cells <- intersect(mc_cells,colnames(mat@mat))
  mc_table <- mat@mat[marker_genes,mc_cells]
  totalumis <- Matrix::colSums(mc_table)
  data.table(total_umis_per_cell = totalumis)
})
umisumsdt <- rbindlist(umisums,idcol = "mc")
umisumsdt[,mc:=factor(mc,levels=mcs)]
# summarize by taking the sum of total UMIs in the cells in each metacell
umisumsdtsum <- umisumsdt[,.(total_umis=sum(total_umis_per_cell)),mc]
```

By combining the previous criteria we identify a pool of potentially bad metacells as those with non-zero UMI counts and total UMI counts below respective distribution peak. 

```{r mc_filt_bad, eval=TRUE}
# combine the metrics
dt <- csumsdtmean[umisumsdtsum,on="mc"]

# non-zero UMI genes threshold
dnonzeroumi <- density(dt$non_zero_genes_peak,adjust=0.4)
nonzerothrs <- 70
plot(dnonzeroumi,main="non-zero UMIs"); abline(v=nonzerothrs)

# total UMI threshold
dumi <- density(dt$total_umis,adjust=0.3)
totumithrs <- 16000
plot(dumi,main="total UMIs"); abline(v=totumithrs)
```

As a final criteria, we retain those potentially bad metacells that express more than 2 known TFs, if they also have moe than 1k UMIs

* **expression of known TFs**

```{r mc_filt_tf, eval=TRUE, fig.width=12, fig.height=3}
TFs_dt <- fread(file.path(annot_dir,"all_tfs_2020-07-08_groups_Xesp_curated.csv"), header=FALSE)
# number of TFs expressed in metacells
tfs <- intersect(TFs_dt$V1,rownames(mc@mc_fp))
tfsmc <- apply(mc@mc_fp[tfs,],2,function(x) sum(x>2.5))
#png(file.path(scdb_fig_dir,sprintf("num_tfs_barplot.png")),width=1800,height=300)
barplot(tfsmc,las=2,xlab="metacell",ylab="number of TFs")
tfthrs <- 5
abline(h=tfthrs)
#dev.off()
```

Inspect metacells' statistics. 

```{r}
tfdt <- data.table(mc=names(tfsmc),num_TFs=tfsmc)
dtt <- dt[tfdt,on="mc"]
# thresholding metacells
dtt[,status:="good"]
dtt[non_zero_genes_peak<nonzerothrs & total_umis<totumithrs, status:="bad"]
dtt[num_TFs>tfthrs & total_umis>1000, status:="good"]
dtt[,num_TFs_scaled:=pmin(num_TFs,5)]
gp <- ggplot(dtt,aes(non_zero_genes_peak,total_umis,label=mc,color=status)) + 
  geom_hline(yintercept=totumithrs, linetype="dashed") +  
  geom_vline(xintercept = nonzerothrs, linetype="dashed") +
  geom_point(aes(size=num_TFs_scaled),alpha=0.9) + scale_size(breaks=1:5,labels=c(1:4,">4"),name="# TFs") +
  geom_text_repel(show.legend=FALSE) + 
  scale_y_log10() + scale_x_log10() +
  labs(x="non-zero UMIs", y="total UMIs") +
  guides(colour=guide_legend(title="metacell",override.aes=list(size=6)))
#ggplotly(gp)
#gp
ggsave(
  file.path(scdb_fig_dir,sprintf("killing_%.0fnonzeroUMI_%.0ftotUMI_%s_TFs.png",nonzerothrs,totumithrs,tfthrs)),
  gp,width=10,height=8
)
```

Remove bad metacells.

```{r mc_kill}
kill_mcs <- dtt[status=="bad",mc]
writeLines(kill_mcs,file.path(scdb_dir,"kill_mcs.txt"))
mc <- mc_set_outlier_mc(mc,as.integer(kill_mcs))
mc <- mc_update_stats(mc,mat)
scdb_add_mc("mc_kill",mc)
```

# Annotation correlations

Load data 

```{r load_data_comp}
mc_id <- ""
mat_id <- ""
mc <- scdb_mc(mc_id)
mat <- scdb_mat(mat_id)
```

Nvec data

```{r load_nvec_comp, eval=TRUE}
old_mcfp <- read.table("/users/asebe/aelek/proj/scRNAseq_nvec_v3/clustering_adult/Xesp_metacell_gene_FC")
old_assign <- fread("/users/asebe/aelek/proj/scRNAseq_nvec_v3/clustering_adult/Xesp_metacell_assignments")
old_annot <- "/users/asebe/aelek/proj/scRNAseq_nvec_v3/clustering_adult/Metacell_annotation_071320"
setnames(old_assign,c("cell","metacell"))
colnames(old_mcfp) <- str_remove(colnames(old_mcfp),"X")
var_genes <- names(which(apply(old_mcfp,1,max)>2))
```

Correlation with Nvec

```{r nvec_comp}
cor_method <- "spearman"
comp_method = "jaccard"
#gene_list <- list("vargenes"=var_genes)
gi <- 1
cors <- sca_clustering_correlation(
  matrix1 = old_mcfp, matrix2 = mc@mc_fp, 
  markers = gene_list[[gi]], 
  original_ordering_1st = TRUE, original_ordering_2nd = TRUE,
  annotation_file_1 = old_annot,
  cor_method = cor_method, pmin = 0.8,
  output_file = file.path(scdb_fig_dir,sprintf("xenia_%s_vs_nvec_%s.png",mc_id,cor_method)),
  height = 1000, width = 1200
)
lfc_threshold <- 1.2
jacc <- sca_clustering_comparison(
  matrix1 = old_mcfp, matrix2 = mc@mc_fp, markers = gene_list[[gi]], lfc_threshold = lfc_threshold,
  original_ordering_1st = TRUE, original_ordering_2nd = TRUE, comp_method = comp_method, 
  annotation_file_1 = old_annot,
  output_file = file.path(
    scdb_fig_dir,sprintf("xenia_%s_vs_nvec_%s_%s_%s.png",mc_id,comp_method,names(gene_list)[gi],lfc_threshold)
  ), height = 2300, width = 2400
)
```

Save cell and metacell annotations.  

```{r, annot_files_save}
mc <- scdb_mc("mc_filt")
mc_annotation <- unique(cell_annotation[,.(mc,cell_type,colour)])
mc_cell_info <- cell_annotation[,.(cell,mc,cell_type,colour)]

mc_annotation_file <- file.path(scdb_dir,"Adult_metacell_annotation")
mc_cell_info_file <- file.path(scdb_dir,"Adult_cell_annotoation")

fwrite(mc_annotation, mc_annotation_file, sep="\t")
fwrite(mc_cell_info, mc_cell_info_file, sep="\t")
```

Fix transferrerd annotations after inspecting 2d plot and heatmap

```{r fix_cols}
mc_annotation_file <- file.path(scdb_dir,"Adult_metacell_annotation_transferred")
mc_annotation <- fread(mc_annotation_file, sep="\t")

# update metacells colors
colid <- match(mc@annots,mc_annotation$mc)
niche_col_new <- mc_annotation[colid,colour]
mc@colors <- niche_col_new

# add new mc object to db
scdb_add_mc(id="mc_filt", mc=mc)
```


# Cell type annotations

Load objects from metacell database

```{r transfer_annots_setup}
mc_id <- "mc_071420"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
```

Save annotation file to use for reordering and annotating metacells. 

```{r mc_annot_template, eval=FALSE}
annk <- fread("clustering_coral/scdb/Metacell_annotation_original_order")
cell_types <- c(
  "Epidermis1","Epidermis2","Epidermis3",
  "Gastrodermis1","Gastrodermis2","Gastrodermis3",
  "Muscle","Cnidocyte","Cnidocyte_precursor","Precursor","Dig_filaments","Gland_secretory",
  paste0("Neuron",1:19), paste0("Unknown",1:5)
)
annk$cell_type <- factor(annk$cell_type,levels=cell_types)
setorder(annk,cell_type,metacell)
fwrite(annk,"clustering_coral/scdb/Metacell_annotation_reord",sep="\t")
```

Create reordered and annotated metacell object.

```{r mc_reord_annot}
annk <- fread("clustering_coral/scdb/Metacell_annotation_071420_reord")

# reorder
mc_reord <- mc_reorder(mc, as.integer(annk$metacell))
mc_reord@colors <- annk$color

# kill
mc_kill <- mc_set_outlier_mc(mc_reord,which(annk$cell_type=="KILL"))
mc_kill <- mc_update_stats(mc_kill,mat)
ann <- annk[cell_type!="KILL"]

# add mc object
scdb_add_mc("mc_083120",mc_kill)

# save final annots
ann_final <- file.path(scdb_dir,"Metacell_annotation_083120")
fwrite(annk[,cell_type!="KILL"][,metacell:=1:.N], ann_final, sep="\t")
```

Save metacell output files.

```{r mc_save}
mc_id <- "mc_083120"
mat_id <- "mat_nowholev2"
mc <- scdb_mc(mc_id)
mat <- scdb_mat(mat_id)
ann_file <- file.path(scdb_dir,"Metacell_annotation_083120")
ann <- fread(ann_file)
fwrite(ann, file.path(scdb_dir,"Xesp_metacell_annotation"), sep="\t")

write.table(
  mc@mc_fp, file=file.path(scdb_dir, "Xesp_metacell_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

mc_umis_list <- lapply(colnames(mc@mc_fp), function(x) {
  cells <- names(mc@mc)[mc@mc==x]
  Matrix::rowSums(mat@mat[,cells])
})
mc_umis <- do.call(cbind,mc_umis_list)
colnames(mc_umis) <- colnames(mc@mc_fp)
mc_umifrac <- mc_umis/rep(colSums(mc_umis),each=nrow(mc_umis))

write.table(
  mc_umis, file=file.path(scdb_dir, "Xesp_metacell_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  mc_umifrac, file=file.path(scdb_dir, "Xesp_metacell_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc@cell_names[!(mc@cell_names %in% mc@outliers)],
  metacell = mc@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Xesp_metacell_assignments"), sep="\t")

saveRDS(mat@mat[,names(mc@mc)], file=file.path(scdb_dir, "Xesp_sc_UMI_counts.RDS"))
```


# Cell type level metacell analysis

Summarize metacells across cell types, and add to the database.

```{r ct_mc}
#source("scr_plot_cmod_markers.R")
mc_id <- "mc_083120"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
ann_file <- file.path(scdb_dir,"Xesp_metacell_annotation")
ann <- fread(ann_file)

# calculate mc object with mc_fp by cell type
mc_ct <- sca_cell_type_fp(
  input_table=ann_file,
  mc_object = mc,
  mat_object = mat
)
scdb_add_mc(id="ct_083120",mc_ct)

# save footprint
write.table(
  mc_ct@mc_fp, file=file.path(scdb_dir, "Xesp_cell_type_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
# save mock ann
ann[,metacell:=cell_type]
#fwrite(unique(ann),file.path(scdb_dir,"Xesp_cell_type_annotation"),sep="\t")

# counts
ct_umis_list <- lapply(colnames(mc_ct@mc_fp), function(y) {
  x <- ann[cell_type==y,]$metacell
  cells <- names(mc_ct@mc)[mc_ct@mc%in%x]
  Matrix::rowSums(mat@mat[,cells])
})
ct_umis <- do.call(cbind,ct_umis_list)
colnames(ct_umis) <- colnames(mc_ct@mc_fp)
ct_umifrac <- ct_umis/rep(colSums(ct_umis),each=nrow(ct_umis))

write.table(
  ct_umis, file=file.path(scdb_dir, "Xesp_cell_type_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  ct_umifrac, file=file.path(scdb_dir, "Xesp_cell_type_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc_ct@cell_names[!(mc_ct@cell_names %in% mc@outliers)],
  cell_type = mc_ct@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Xesp_cell_type_assignments"), sep="\t")

```

# Broad cell type level metacell analysis

Group  cell types into broad cell types, and add to the database.

```{r bct_mc}
#source("scr_plot_cmod_markers.R")
mc_id <- "ct_083120"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)
ann_file <- file.path(scdb_dir,"Xesp_broad_cell_type_annotation")
ann <- fread(ann_file)

# calculate mc object with mc_fp by cell type
mc_ct <- scr_cell_type_fp(
  input_table = ann_file,
  mc_object = mc,
  mat_object = mat
)
scdb_add_mc(id="bct_083120",mc_ct)

# save footprint
write.table(
  mc_ct@mc_fp, file=file.path(scdb_dir, "Xesp_broad_cell_type_gene_FC"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

# save mock ann
ann[,metacell:=cell_type]
#fwrite(unique(ann),file.path(scdb_dir,"Xesp_broad_cell_type_annotation"),sep="\t")

# counts
ct_umis_list <- lapply(colnames(mc_ct@mc_fp), function(y) {
  x <- ann[cell_type==y,]$metacell
  cells <- names(mc_ct@mc)[mc_ct@mc%in%x]
  Matrix::rowSums(mat@mat[,cells])
})
ct_umis <- do.call(cbind,ct_umis_list)
colnames(ct_umis) <- colnames(mc_ct@mc_fp)
ct_umifrac <- ct_umis/rep(colSums(ct_umis),each=nrow(ct_umis))

write.table(
  ct_umis, file=file.path(scdb_dir, "Xesp_broad_cell_type_UMI_counts"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)
write.table(
  ct_umifrac, file=file.path(scdb_dir, "Xesp_broad_cell_type_UMI_fractions"), 
  sep="\t", quote=FALSE, row.names=TRUE, col.names=TRUE
)

cell_df <- data.table(
  cell = mc_ct@cell_names[!(mc_ct@cell_names %in% mc@outliers)],
  cell_type = mc_ct@mc
)
fwrite(cell_df, file=file.path(scdb_dir, "Xesp_broad_cell_type_assignments"), sep="\t")

```

# Sequencing and clustering statistics

Load objects from metacell database

```{r}
mc_id <- "mc_083120"
mat_id <- "mat_nowholev2"
mat <- scdb_mat(mat_id)
mc <- scdb_mc(mc_id)

bl <- readLines("annotation_xenia/xenia_bl")

ann <- fread(file.path(scdb_dir,"Xesp_metacell_annotation"))
setnames(ann,c("metacell","cell_type","color"))
niche_order <- as.character(ann$metacell)
niche_color <- ann$color
```

Extract stats for metacell clustering.

```{r}
# cell sizes
cell_sizes <- apply(mat@mat,2,sum)
cdt <- data.table(cells=colnames(mat@mat),cell_sizes=cell_sizes)
cdt[data.table(cells=names(mc@mc),metacells=mc@mc), on="cells", mc:=i.metacells]
cdt <- cdt[!is.na(mc)]
cdt[,median_mc_cell_size:=median(cell_sizes),mc]
cdt[,mc:=factor(mc,levels=sort(as.integer(unique(cdt$mc))))]
cdt[,total_mc_umis:=sum(cell_sizes),mc]
cdt[,total_mc_cells:=.N,mc]
fwrite(cdt,file.path(scdb_dir,"cdt.tsv"),sep="\t")
# gene sizes
gene_umis <- apply(mat@mat,1,sum)
gdt <- data.table(gene=rownames(mat@mat),gene_umis=gene_umis)
fwrite(gdt,file.path(scdb_dir,"gdt.tsv"),sep="\t")
# gene stats
gstat_id <- "gstat_nowholev2"
gstat <- scdb_gstat(gstat_id)
gset_id <- "clust_markers_nowholev2_bl"
marker_genes_gset <- scdb_gset(gset_id)
markers <- names(marker_genes_gset@gene_set)
gstat <- setDT(gstat)
setnames(gstat,"name","gene")
gstat[, marker:="no"]
gstat[gene %in% markers, marker:="yes"]
gstat[,log2_ds_mean:=log2(ds_mean)]
fwrite(gstat,file.path(scdb_dir,"gstat.tsv"),sep="\t")
# markers in cells
markers_mat <- mat@mat[markers,cdt$cells]
markers_num <- apply(markers_mat, 2, function(x) sum(x>0))
mdf <- cbind.data.frame(num_markers=markers_num)
fwrite(mdf,file.path(scdb_dir,"mdf.tsv"),sep="\t")
```

Count reads from CellRanger output bam files.

```{bash}
#!/bin/bash
#$ -V
#$ -cwd
#$ -N bamcount
#$ -M anamaria.elek@crg.eu
#$ -m a
#$ -o read_counts.out
#$ -j y
#$ -q long-sl7,mem_512_12h
#$ -l virtual_free=20G,h_rt=43200
#$ -pe smp 12

cores=4
ml purge
ml BEDTools/2.27.1-foss-2016b

BAMS=( ls *bam )
COUNTS="/clustering_xenia/scdb/read_counts.csv"
echo file,total_reads,mapped_reads > $COUNTS
for bam in $BAMS
do
  echo $bam
  countsc=${bam%%bam}.CB.counts.txt
  countscmq=${bam%%bam}mapq30.CB.counts.txt
  echo "Counting BCs"
  samtools view $bam | grep CB:Z: | sed 's/.*CB:Z:\([ACGT]*\).*/\1/' | sort | uniq -c > $countsc
  echo "Counting BCs for mapq30"
  samtools view -c -q 30 -F 260 $bam | grep CB:Z: | sed 's/.*CB:Z:\([ACGT]*\).*/\1/' | sort | uniq -c > $countscmq
  # TOTAL_READS=$( samtools view -c $bam )
  # MAPPED_READS=$( samtools view -c -q 30 -F 260 $bam )
  # echo ${bam},${TOTAL_READS},${MAPPED_READS} >> $COUNTS
done

```

Summarize

```{r}

cdt <- fread(file.path(scdb_dir,"cdt.tsv"))
cdt[,dataset:=str_extract(cells,'(tentacle|whole|stalk)_v\\d+')]
dt <- merge.data.table(
  cdt[,.(cells=.N),dataset],
  cdt[,.(UMI_per_cell=median(cell_sizes)),dataset],
  by="dataset"
)

counts <- rbindlist(lapply(list.files(scdb_dir,"read.counts.csv",full.names=TRUE)[-1],fread))
counts[,dataset:=str_extract(file,'(tentacle|whole|stalk)_v\\d+')][,file:=NULL]
stats <- merge.data.table(dt,counts,by="dataset",all.x=TRUE)
stats[,total_reads_per_cell:=total_reads/cells]
stats[,mapped_reads_per_cell:=mapped_reads/cells]
stats[,mapped_reads_per_UMI:=mapped_reads_per_cell/UMI_per_cell]
fwrite(stats,file.path(scdb_dir,"stats.tsv"),sep="\t")

```

# Session info

```{r session, eval=TRUE}
sessionInfo()
```

